straxen/common.py

```diff
@@ -243,7 +247,185 @@ def get_livetime_sec(context, run_id, things=None):
+def remap_channels(data, verbose=True, safe_copy=False, _tqdm=False, ):
+    """
+    There were some errors in the channel mapping of old data as described in
+        https://xe1t-wiki.lngs.infn.it/doku.php?id=xenon:xenonnt:dsg:daq:sector_swap
+        using this function, we can convert old data to reflect the right channel map
+        while loading the data. We convert both the field 'channel' as well as anything
+        that is an array of the same length of the number of channels.
+
+    :param data: numpy array of pandas dataframe
+
+    :param verbose: print messages while converting data
+
+    :param safe_copy: if True make a copy of the data prior to performing manipulations.
+        Will prevent overwrites of the internal references but does require more memory.
+    
+    :param _tqdm: bool (try to) add a tqdm wrapper to show the progress
+
+    :return: Correctly mapped data
+    """
+    # This map shows which channels were recabled. We now have to do the same in software
+    # for old runs.
+    remap = get_resource(
+        aux_repo + '/ecb6da7bd4deb98cd0a4e83b3da81c1e67505b16/remapped_channels_since_20200729_17.20UTC.csv',
+        fmt='csv')
+
+    def wr_tqdm(x):
+        """Wrap input x with tqdm"""
+        if _tqdm:
+            try:
+                return tqdm.tqdm_notebook(x)
+            except (AttributeError, ModuleNotFoundError, ImportError):
+                # ok, sorry lets not wrap but return x
+                pass
+        return x
+
+    def get_dtypes(_data):
+        """
+        Return keys/dtype names of pd.DataFrame or numpy array
+
+        :param _data: data to get the keys/dtype names
+
+        :return: keys/dtype names
+        """
+        if isinstance(_data, np.ndarray):
+            _k = _data.dtype.names
+        elif isinstance(_data, pd.DataFrame):
+            _k = _data.keys()
+        return _k
+
+    def convert_channel(_data, replace=('channel', 'max_pmt')):
+        """
+        Given an array, replace the 'channel' entry if we had to remap it according to the
+            map of channels to be remapped
+
+        :param _data: data whereof the replace entries should be changed according to the
+            remapping of channels
+
+        :param replace: entries (keys/numpy-dtypes) that should be changed in the data
+
+        :return: remapped data where each of the replace entries has been replaced
+        """
+        # loop over the things to replace
+        for _rep in replace:
+            _k = get_dtypes(_data)
+            if _rep not in _k:
+                # Apparently this data doesn't have the entry we want to replace
+                continue
+            # Make a buffer we can overwrite and replace with an remapped array
+            buff = np.array(_data[_rep])
+            _n_replacements = 0
+            for _, _row in wr_tqdm(remap.iterrows()):
+                pmt_new, pmt_old = _row['PMT_new'], _row['PMT_old']
+                mask = _data[_rep] == pmt_old
```

Comment:
Boolean indexing is super slow. Especially on large arrays (if I remember correctly). Maybe it would be better to change this in a simple numba function?

Reply:
Thanks, indeed you are right:
```python
%%timeit
convert_channel(rr)
```
The implementation above:
3.28 s ± 14.3 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)

With Numba (after first initiation):
568 ms ± 10.2 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)